LiteCore/tests/QueryTest.cc

```diff
@@ -336,6 +338,44 @@ N_WAY_TEST_CASE_METHOD(QueryTest, "Query boolean", "[Query]") {
     }
 }
 
+N_WAY_TEST_CASE_METHOD(QueryTest, "Query uint64", "[Query]") {
+    {
+        ExclusiveTransaction t(store->dataFile());
+        const int64_t ivals[] = { std::numeric_limits<int64_t>::min(),
+                                  std::numeric_limits<int64_t>::max() };
+        const uint64_t uvals[] = { 0, std::numeric_limits<uint64_t>::max(), };
+        for(int i = 0; i < 4; i++) {
+            string docID = stringWithFormat("rec-%03d", i + 1);
+            writeDoc(slice(docID), DocumentFlags::kNone, t, [=](Encoder &enc) {
+                enc.writeKey("value");
+                if (i < 2)
+                    enc.writeInt(ivals[i]);
+                else
+                    enc.writeUInt(uvals[i - 2]);
+            });
+        }
+        t.commit();
+    }
+
+    // Check the data type of the returned values:
+    Retained<Query> query = store->compileQuery(json5( "{WHAT: ['.value']}"));
+    Retained<QueryEnumerator> e = query->createEnumerator();
+    REQUIRE(e->getRowCount() == 4);
+    int row = 0;
+    while (e->next()) {
+        auto type = e->columns()[0]->type();
+        CHECK(type == kNumber);
+        auto isInt = e->columns()[0]->isInteger();
+        CHECK(isInt);
+        auto isUnsigned = e->columns()[0]->isUnsigned();
+        if (row < 3)
```

Comment:
Should this be "row < 2" ?

Reply:
No, this is special case < 2048:
```
 void Encoder::writeUInt(uint64_t i) {writeInt(i, (i < 2048),               true);} 
 void Encoder::writeInt(uint64_t i, bool isSmall, bool isUnsigned) {                                                                                                         
        if (isSmall) {                                                                                                                                                          
            new (placeItem()) Value(kShortIntTag, (i >> 8) & 0x0F, i & 0xFF);   
        } else
```
so during fleece encoding we lost all information about that it was unsigned.

Reply:
I see. I read the relevant code over the weekend and I think, first of all, your fix in fleece function is definitely right. Second, whether to encode in fleece as signed or unsigned is an implementation decision with the constraint of being correct representing the number. With 64-bit int, any number whose absolute value is less than or equal to numeric_limit<int64_t>::max can be represented either in signed or unsigned formats. SQLite stopped here; they only use signed int. LiteCore uses extra bit to allow greater numbers. That's why, even though you call enc.wirteUInt, we may still encode it in singed way because its scale can fit in our signed representation. The actual cutoff now is 12 bits signed (absolute value fits in 11 bits.)
Therefore, the meaningful checks would be:
* for negative number, its fleece representation must be signed. Or
      ivals[r] >= 0 || !col->isUnsiigned()
* for positive number, its fleece representation unsigned if it cannot fit into 63 bits. Or
     col->asInt() >= 0 || col->isUnsigned()
That is, for negative numbers, isUnsigned() == false regardless the magnitude. For positive numbers, even you call enc->writeUInt(), we may still represent it in signed format provided its magnitude allows faithful representation as numbers.
Therefore, I would suggest that test focus on whether Fleece correctly represent the number correctly, not to use the knowledge based on the above code of small integer.